File: Lesson.02 - Design Patterns - Creational/FactoryMethod/src/main/java/com/github/ASE/FactoryMethod/PlainText/DocumentGenerator.java
package com.github.ASE.FactoryMethod.PlainText;

import com.github.ASE.FactoryMethod.Footer;
import com.github.ASE.FactoryMethod.Header;
import com.github.ASE.FactoryMethod.Paragraph;
import com.github.ASE.FactoryMethod.Table;

public class DocumentGenerator extends com.github.ASE.FactoryMethod.DocumentGenerator {

    @Override
    public Header createHeader() {
        return new Header() {
            @Override
            public String getContent() {
                return "=== Document Title ===";
            }
        };
    }

    @Override
    public Paragraph createParagraph() {
        return new Paragraph() {
            @Override
            public String getContent() {
                return "This is a plain text paragraph.";
            }
        };
    }

    @Override
    public Table createTable() {
        return new Table() {
            @Override
            public String getContent() {
                return "Name | Age\n-----|-----\nJohn | 30";
            }
        };
    }

    @Override
    public Footer createFooter() {
        return new Footer() {
            @Override
            public String getContent() {
                return "--- End of Document ---";
            }
        };
    }

}

File: Lesson.02 - Design Patterns - Creational/Python/FactoryMethod/plaintext/document_generator.py
from document_generator import DocumentGenerator
from components.header import Header
from components.paragraph import Paragraph
from components.table import Table
from components.footer import Footer


class PlainTextDocumentGenerator(DocumentGenerator):
    def create_header(self) -> Header:
        return PlainTextHeader()

    def create_paragraph(self) -> Paragraph:
        return PlainTextParagraph()

    def create_table(self) -> Table:
        return PlainTextTable()

    def create_footer(self) -> Footer:
        return PlainTextFooter()


class PlainTextHeader(Header):
    def get_content(self) -> str:
        return "=== Document Title ==="


class PlainTextParagraph(Paragraph):
    def get_content(self) -> str:
        return "This is a plain text paragraph."


class PlainTextTable(Table):
    def get_content(self) -> str:
        return "Name | Age\n-----|-----\nJohn | 30"


class PlainTextFooter(Footer):
    def get_content(self) -> str:
        return "--- End of Document ---"
